wikidata/search.py

```diff
@@ -0,0 +1,415 @@
+import html.parser
+import re
+import urllib.error
+import urllib.parse
+import urllib.request
+from dataclasses import dataclass
+from difflib import SequenceMatcher, get_close_matches
+from typing import List, Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .client import Client
+    from urllib.request import OpenerDirector
+else:
+    WIKIDATA_BASE_URL = 'https://www.wikidata.org/'
+    
+from .entity import EntityId
+
+__all__ = ['SearchResult', 'suggest_query', 'filter_by_similarity']
+
+
+@dataclass 
+class SearchResult:
+    """Represents a single search result from Wikidata.
+    
+    :param entity_id: The Wikidata entity ID (e.g., 'Q790')
+    :type entity_id: :class:`EntityId`
+    :param name: The official name of the entity
+    :type name: :class:`str`
+    :param similarity_score: The similarity score between the search query and the result name
+                           as a percentage (0-100)
+    :type similarity_score: :class:`float`
+    :param description: Optional description of the entity
+    :type description: :class:`str` or :const:`None`
+    """
+    entity_id: EntityId
+    name: str
+    similarity_score: float
+    description: Optional[str] = None
+
+    def __repr__(self) -> str:
+        return (
+            f'SearchResult(entity_id={self.entity_id!r}, name={self.name!r}, '
+            f'similarity_score={self.similarity_score:.1f}%, '
+            f'description={self.description!r})'
+        )
+
+def _search(
+    query: str,
+    limit: int = 10,
+    opener: Optional['OpenerDirector'] = None,
+    client: Optional['Client'] = None,
```

Comment:
It accepts both `opener` and `client` parameters. Since `client` already contains `opener`, consider simplifying to accept only `client` (or make it a method of `Client`).